rust/rbac-registration/src/cardano/cip509/cip509.rs

```diff
@@ -0,0 +1,572 @@
+//! Cardano Improvement Proposal 509 (CIP-509) metadata module.
+//! Doc Reference: <https://github.com/input-output-hk/catalyst-CIPs/tree/x509-envelope-metadata/CIP-XXXX>
+//! CDDL Reference: <https://github.com/input-output-hk/catalyst-CIPs/blob/x509-envelope-metadata/CIP-XXXX/x509-envelope.cddl>
+
+use std::{borrow::Cow, collections::HashMap};
+
+use anyhow::{anyhow, Context};
+use cardano_blockchain_types::{MetadatumLabel, MultiEraBlock, TxnIndex};
+use catalyst_types::{
+    hashes::{Blake2b256Hash, BLAKE_2B256_SIZE},
+    problem_report::ProblemReport,
+};
+use cbork_utils::decode_helper::{decode_bytes, decode_helper, decode_map_len};
+use minicbor::{
+    decode::{self},
+    Decode, Decoder,
+};
+use pallas::{
+    codec::utils::Nullable,
+    ledger::{
+        addresses::{Address, ShelleyAddress},
+        primitives::conway,
+        traverse::MultiEraTx,
+    },
+};
+use strum_macros::FromRepr;
+use tracing::warn;
+use uuid::Uuid;
+
+use crate::{
+    cardano::cip509::{
+        decode_context::DecodeContext,
+        rbac::Cip509RbacMetadata,
+        types::{PaymentHistory, RoleNumber, TxInputHash, ValidationSignature},
+        utils::Cip0134UriSet,
+        validation::{
+            validate_aux, validate_role_data, validate_stake_public_key, validate_txn_inputs_hash,
+        },
+        x509_chunks::X509Chunks,
+        Payment, PointTxnIdx, RoleData,
+    },
+    utils::decode_helper::{report_duplicated_key, report_missing_keys},
+};
+
+/// A x509 metadata envelope.
+///
+/// The envelope is required to prevent replayability attacks. See [this document] for
+/// more details.
+///
+/// [this document]: https://github.com/input-output-hk/catalyst-CIPs/blob/x509-envelope-metadata/CIP-XXXX/README.md
+#[derive(Debug, Clone)]
+#[allow(clippy::module_name_repetitions)]
+pub struct Cip509 {
+    /// A registration purpose (`UUIDv4`).
+    ///
+    /// The purpose is defined by the consuming dApp.
+    purpose: Option<Uuid>,
```

Comment:
I think it could you use a `UuidV4` type from the `catalyst-types` crate, so we will be totally sure that its a uuid v4.